Better handling of public suffixes during domain creation

peterthomassen opened the pull request:
wip - details follow

peterthomassen commented:
(Sorry for the initial useless PR message.)

This PR does three things It

1. moves the special role of `dedyn.io` from the application logic to a setting,
2. catches duplicate domain creation using a trial-and-error approach (save and error out),
3. checks the Public Suffix List and denies domain creation for public suffixes. The lookup is done via a DNS service using the [psl-dns library](https://pypi.org/project/psl-dns/). I think this is pretty failsafe, especially as there are no hard consequences in case of lookup failure.

Regarding the PSL-related tests, they currently run with the real DNS lookups. It's [easily possible to mock them](https://stackoverflow.com/a/51707059/6867099), but running the DNS requests in the tests has the subtle benefit that we'll have a few extra insights into the reliability of the service. I'm open to discussions though, including using other means of looking up PSL rules if they turn out to be easier / safer / faster / ...

nils-wisiol commented:
1. Suffix management:
    - there are one or two PEP8 violations
    - What do we expect if we directly create a domain `a.b.dedyn.io`? Currently, no auto delegation takes place as a test that I added (dba8d0e821c6c4af598fd957813fc6e5ef87fbdc) confirms. (*)
    - Should we avoid the code duplication in the suffix check?

          subname, parent_domain = self.name.split('.', 1)
          if parent_domain in settings.LOCAL_PUBLIC_SUFFIXES:

      It could be replaced with an object method like `if self.has_public_suffix_parent()` or similar, depending on the answert to (*). Even more concise, it could be replaced with `self.get_public_suffix_parent()` to avoid the following try/catch.
    - otherwise, lgtm

2. Domain creation exception handling:
    - should there be an explicit test for each case?
    - otherwise, lgtm

3. PSL check:
    - the check should be moved to an [object level permission](https://www.django-rest-framework.org/api-guide/permissions/#examples) (as should, eventually, all other checks in `perform_create`)
    - this introduces a dependency on un-reviewed code. While we do not review dependency code, it also doesn't make sense to exclude home-brew code from review just because it lives somewhere else.
    - We're querying our own database via our own DNS servers, using the Internet. Hence, the check at hand will rely on unrelated business, such as DNS load balancers, DDoS mitigation, front end configuration, and so on. But the data is right there, in `Domain.objects`. The scenario where we rely on someone else's PSL hosting service seems unlikely, as this code is, while being a library with pip integration, still written by you and also does only implement a subset of the PSL specification.
    - The tests add a Internet-connection dependency on the test code that was just cleaned up and became independent of any connections to the outside world. The tests fail without Internet connection by design.
    - When there's no Internet connection, https://github.com/desec-io/desec-stack/pull/184/commits/0ef595fccca03051295e95565f47d8f7657619c3#diff-20b11db83f0d07f65b8705bd47502d64R155 either raises an exception or returns status code 500 (It's hard to debug because you can't google while debugging; what I saw was status 500 in the debugger but also an unhandled exception before the status was checked in the following line (so the test errored and did not fail). Which status code should we expect in that case?)

For number 3, I believe you already invested too much work to withdraw this PR. So let's fix the most pressing issues and then merge and move on.

peterthomassen commented:
> 1. Suffix management:
>    
>    * there are one or two PEP8 violations

Which?

>    * What do we expect if we directly create a domain `a.b.dedyn.io`? Currently, no auto delegation takes place as a test that I added (dba8d0e821c6c4af598fd957813fc6e5ef87fbdc) confirms. (*)

In this case, no autodelegation should happen, because somebody else could come and register `b.dedyn.io`, and would not be aware of the delegation.

>    * Should we avoid the code duplication in the suffix check?
>      ```
>      subname, parent_domain = self.name.split('.', 1)
>      if parent_domain in settings.LOCAL_PUBLIC_SUFFIXES:
>      ```
>      
>      
>      It could be replaced with an object method like `if self.has_public_suffix_parent()` or similar, depending on the answert to (*). Even more concise, it could be replaced with `self.get_public_suffix_parent()` to avoid the following try/catch.

I thought about this for a few minutes now, but unfortunately we also need the `subname` in the payload, so that `split()` call needs to be in the same scope. I couldn't come up with an elegant abstraction, so I'd leave it as is for now.

>    * please shorten commit message title

Done

> 2. Domain creation exception handling:
>    
>    * should there be an explicit test for each case?

We already test registering domains when the same domain is already registered, or when a parent is registered by another user. The new code is such that the first case is caught by the database uniqueness check (so we have coverage).

> 3. PSL check:
>    
>    * the check should be moved to an [object level permission](https://www.django-rest-framework.org/api-guide/permissions/#examples) (as should, eventually, all other checks in `perform_create`)

My understanding of permissions is that depending on the request, the outcome may be "allow" or "deny" access to an existing resource. Here, the resource does not exist yet. Do you know whether permissions can also be used at creation time? (I haven't found such an example.)

>    * this introduces a dependency on un-reviewed code. While we do not review dependency code, it also doesn't make sense to exclude home-brew code from review just because it lives somewhere else.

I agree fully. Shall we review it? It's quite long, but I think it would be almost equally long if it was part of `desec-stack`, at least as far as the PSL parser is concerned.

>    * We're querying our own database via our own DNS servers, using the Internet. Hence, the check at hand will rely on unrelated business, such as DNS load balancers, DDoS mitigation, front end configuration, and so on. But the data is right there, in `Domain.objects`. The scenario where we rely on someone else's PSL hosting service seems unlikely, as this code is, while being a library with pip integration, still written by you and also does only implement a subset of the PSL specification.

If others will use our stack (and our upcoming sponsor will), they will most likely use our PSL query service, or the other way around (no reason to keep two PSL zones in sync). I would consider this service external, and in our case, the service operator just happens to use our DNS hosting platform.

Regarding failure modes like DDoS etc, I agree that we need to be resilient. As the PSL is recommended to be updated no more often than daily, occasional out-of-sync is acceptable by design of the PSL. So, my stand is that we can be permissive if the lookup can't be done.

Also, it can happen that users already have domains which are under a domain that gets added to the PSL later on. To catch these cases, we can set up a cronjob or something that notifies us so that we can intervene, if necessary. This same cronjob would also catch cases in which we are permissive because of lookup failure.

Regarding the subset: The missing part is currently not used in the PSL itself. Also, I thought up a way how to fix this (we can discuss in person). However, I'd first like to try changing the specs because I consider that part quite misdirected.

>    * The tests add a Internet-connection dependency on the test code that was just cleaned up and became independent of any connections to the outside world. The tests fail without Internet connection by design.

Yep. Do we agree that the way forward is to mock those?

>    * When there's no Internet connection, [0ef595f#diff-20b11db83f0d07f65b8705bd47502d64R155](https://github.com/desec-io/desec-stack/commit/0ef595fccca03051295e95565f47d8f7657619c3#diff-20b11db83f0d07f65b8705bd47502d64R155) either raises an exception or returns status code 500 (It's hard to debug because you can't google while debugging; what I saw was status 500 in the debugger but also an unhandled exception before the status was checked in the following line (so the test errored and did not fail). Which status code should we expect in that case?)

I'll look into it.

> For number 3, I believe you already invested too much work to withdraw this PR. So let's fix the most pressing issues and then merge and move on.

It is completely fine with me to withdraw this if we find a superior solution. I just think this query service is a good idea to not overload out stack, and the failure modes can be dealt with. I'm completely non-emotional about doing something else and throwing this away if the solution is not convincing.

nils-wisiol commented:
Yes, let's mock the DNS replies for the sake of the tests. Also, the dependency should be upgraded to a stable version before launch (but not necessarily before this can be merged). Yes, we should review and/or test the psl code base. Let's save the cleaning up of the permissions check for a later PR, as it seems quite involved, and is really more of a clean-up task than a feature implementation (like this PR). After addressing those items, I'm okay with merging this

peterthomassen commented:
I added mocking of the PSL library to all tests; all tests now pass in offline mode. That was quite an adventure, but I learned a lot about `unittest.mock`. Very interesting stuff! (May also be applicable to pdns mocking, for example, by intercepting calls of the functions in `pdns.*`, in case we get tired of dealing with httpretty at some point in the future ...)

I'm planning to add tests to the library itself and then promote it to a stable version before launch, yes. It's probably best to review the code base once tests are there (but feel free to start if you're interested).

I also added a test that fakes an `UnsupportedRule` exception during PSL lookup and makes sure that domain creation can proceed in this case. (As mentioned earlier, this is purely hypothetical now.) After doing that, I realized that we could also just not catch the exception in `perform_create()`, and instead add an extra two lines [here](https://github.com/desec-io/desec-stack/blob/f807e9cfbbbb105cb9aa80908946868e985b6793/api/desecapi/exception_handlers.py#L39) so that the user gets `503 Service Unavailable` and we receive an email. This way, the situation would not go unnoticed. Do you think we should go that way? (I think that it probably makes no difference in practice as we'll have to check existing domains against PSL updates anyways every once in a while.)

With the above done, I'm not aware of any outstanding items. Once you reviewed again, I can squash the new commits and we can proceed.

peterthomassen commented:
fyi, I created tests for the imported PSL library code that is used in this PR: https://github.com/peterthomassen/psl-dns/commit/c2113c0346da4ab3003b41a1ea2f00488f56735e

nils-wisiol commented:
> This way, the situation would not go unnoticed. Do you think we should go that way?

I don't really care, but since we believe it doesn't happy anyway I'm okay with planning for a 503 and an email

nils-wisiol commented on api/desecapi/views.py:
```diff
@@ -133,16 +135,15 @@ def perform_create(self, serializer):
+                "code": "domain-unavailable"}
+            )
+            ex.status_code = status.HTTP_409_CONFLICT
+            raise ex
```
Just as a general note: While refactoring for #97, I found that there is at least one more type how things can go wrong, which is a Django `ValidationError`. I hence believe there are in total four ways that need to be considered when creating objects (most notably Domains and RRSets):

1. IntegrityError coming from the api's database
2. PdnsException coming from, well, powerdns
3. Django's ValidationError coming from the model layer
4. DRF's ValidationError coming from the serializer

I'm not saying all of them are needed o be caught here, but just wanted to note this for future reference.

nils-wisiol commented on api/desecapi/tests/test_domains.py:
```diff
@@ -163,6 +164,14 @@ def test_create_under_public_suffix_with_private_parent(self):
+    def test_create_domain_under_unsupported_public_suffix_rule(self):
+        # Show lenience if the PSL library produces an UnsupportedRule exception
+        name = 'unsupported.wildcard.test'
+        psl_cm = self.get_psl_context_manager(UnsupportedRule)
```
If I understand correctly, this passes an `Exception` type object to the `public_suffix` argument in `_mock_get_public_suffix`. That this actually works is quite surprising, consider renaming the argument of that function

peterthomassen replied:
The code passes a class (derived from `Exception`) to `get_psl_context_manager`. The class is callable (calling it produces the instance). `get_psl_context_manager` checks for this case, and if the thing at hand is callable, it does not pass it on to `_mock_get_public_suffix`, but instead uses it directly to set the `side_effect` attribute of the mocked function.

Side effects in turn are one of the neat features of `unittest.mock`. You can set a `side_effect` attribute on your mocked objects/methods/... which takes effect (haha) when the mock is called. They have logic that if the side effect is an `Exception` instance, it is raised and not called. That's actually pretty cool, because you can quickly simulate various error scenarios. https://docs.python.org/3/library/unittest.mock.html#quick-guide

One could pass any other callable (most useful would be a function) to the context manager. The function then has to either produce a return value for the mocked method, or raise an exception. I was amazed how flexible the mock library is.

Because of the flexibility, it's hard to find a good name for that argument, so I made a bad choice :) I changed it to `side_effect_parameter`.

peterthomassen commented on api/desecapi/tests/test_domains.py:
```diff
@@ -157,18 +182,20 @@ def test_create_domain_atomicity(self):
             self.assertFalse(Domain.objects.filter(name=name).exists())
 
     def test_create_domain_punycode(self):
```
Actually, we have separate public suffix tests, so this should not be here. I reverted the change to the test. (It was an artifact from developing the PSL stuff before mocking the PSL test calls; that made this test break as one of the test domain names is on the PSL. With mocking, it's not an issue anymore.)

peterthomassen commented on api/desecapi/views.py:
```diff
@@ -101,25 +102,41 @@ def get_queryset(self):
-        queryset = Domain.objects.filter(Q(name=domain_name) | (Q(name__in=domain_list) & ~Q(owner=self.request.user)))
-        if queryset.exists():
+        # Check if domain is a public suffix
+        psl = psl_dns.PSL(resolver=settings.PSL_RESOLVER)
```
done

peterthomassen replied:
In fact, moving this requires adapting the mocking mechanism because now the `psl` object already exists when the mock is initialized, i.e. we have to patch the `psl` object and not the `PSL` class. Not a big deal, just to explain the changes.

peterthomassen commented on api/desecapi/views.py:
```diff
@@ -101,25 +102,41 @@ def get_queryset(self):
+            is_public_suffix = psl.is_public_suffix(domain_name)
+        except psl_dns.exceptions.UnsupportedRule:
+            # Don't be too restrictive if indeterminate
+            public_suffix = domain_name.rsplit('.')[-1]  # `*` PSL rule always applies
```
Done. - The `maxsplit` argument is actually of no consequence (unless we set it to `0`).

peterthomassen replied:
Update: This code went away because I changed error handling such that we reraise the exception in order to produce a 503 along with an email. The view now exits here.

peterthomassen commented on api/desecapi/views.py:
```diff
@@ -101,25 +102,41 @@ def get_queryset(self):
+        # If another user owns a zone with one of these names, then the requested
+        # domain is unavailable because it is part of the other user's zone.
+        private_components = domain_name.rsplit(public_suffix, 1)[0].rstrip('.')
+        private_components = private_components.split('.') if private_components else []
```
nope: empty string split arbitrarily results in `['']`, not in `[]`

peterthomassen commented:
Thanks for all the feedback. I pushed one more commit that takes it into account. Can you please take a look at it as well? - I'll then squash the PSL functionality into one commit (no need to have the development history in master).
